internal/webhook/external/azure.go

```diff
@@ -0,0 +1,219 @@
+	secretData map[string][]byte,
+) ([]string, error) {
+	secretValue, ok := secretData[azureSecretDataKey]
+	if !ok {
+		return nil,
+			fmt.Errorf("missing data key %q for Azure WebhookReceiver", azureSecretDataKey)
+	}
+	return []string{string(secretValue)}, nil
+}
+
+// getHandler implements WebhookReceiver.
+func (a *azureWebhookReceiver) getHandler(requestBody []byte) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("received azure event", "userAgent", r.UserAgent())
+
+		switch {
+		// format is AzureContainerRegistry/<version>
+		case strings.Contains(r.UserAgent(), "AzureContainerRegistry"):
+			a.handleAcrEvent(ctx, w, requestBody)
+		// Format is VSServices/<version>
+		case strings.Contains(r.UserAgent(), "VSServices"):
+			a.handleAzureDevOpsEvent(ctx, w, requestBody)
+		default:
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					errors.New(
+						"request does not appear to have originated from a supported service",
+					),
+					http.StatusBadRequest,
+				),
+			)
+			return
+		}
+	})
+}
+
+func (a *azureWebhookReceiver) handleAcrEvent(
+	ctx context.Context,
+	w http.ResponseWriter,
+	requestBody []byte,
+) {
+	var event acrEvent
+	if err := json.Unmarshal(requestBody, &event); err != nil {
+		xhttp.WriteErrorJSON(
+			w,
+			xhttp.Error(errors.New("invalid request body"),
+				http.StatusBadRequest,
+			),
+		)
+		return
+	}
+
+	switch event.Action {
+	case "push":
+		repoURLs := resolveAcrRepoURLs(event)
+		logger := logging.LoggerFromContext(ctx)
+		logger = logger.WithValues("repoURLs", repoURLs)
+		ctx = logging.ContextWithLogger(ctx, logger)
+		refreshWarehouses(ctx, w, a.client, a.project, repoURLs...)
+	case "ping":
+		xhttp.WriteResponseJSON(
+			w,
+			http.StatusOK,
+			map[string]string{
+				"msg": "ping event received, webhook is configured correctly",
+			},
+		)
+	default:
+		xhttp.WriteErrorJSON(
+			w,
+			xhttp.Error(
+				fmt.Errorf("event type %s is not supported", event.Action),
+				http.StatusBadRequest,
+			),
+		)
+		return
+	}
+}
+
+func (a *azureWebhookReceiver) handleAzureDevOpsEvent(
+	ctx context.Context,
+	w http.ResponseWriter,
+	requestBody []byte,
+) {
+	var event azureDevOpsEvent
+	if err := json.Unmarshal(requestBody, &event); err != nil {
+		xhttp.WriteErrorJSON(
+			w,
+			xhttp.Error(errors.New("invalid request body"),
+				http.StatusBadRequest,
+			),
+		)
+		return
+	}
+	repoURL := git.NormalizeURL(event.Resource.Repository.RemoteURL)
+	logger := logging.LoggerFromContext(ctx)
+	logger = logger.WithValues("repoURL", repoURL)
+	ctx = logging.ContextWithLogger(ctx, logger)
+	refreshWarehouses(ctx, w, a.client, a.project, repoURL)
+}
+
+func resolveAcrRepoURLs(e acrEvent) []string {
+	repoURL := fmt.Sprintf("%s/%s", e.Request.Host, e.Target.Repository)
+	var repoURLs []string
+	switch e.Target.MediaType {
+	case imageMediaType:
```

Comment:
We should probably also include `application/vnd.oci.image.manifest.v1+json` in this case.

There are existing receivers where we should fix that and I volunteer to do so.

Reply:
Ah... there are even more. Just ignore this bit of feedback and I will write a function that checks for all valid image media types.